virtualbox/vboxcommon.py

```diff
@@ -101,72 +101,50 @@ def get_vm_state(vm_uuid):
     raise Exception(f"Could not start VM '{vm_uuid}'")
 
 
-def ensure_vm_running(vm_uuid):
-    """Checks if the VM is running and starts it if it's not.
-    Waits up to 1 minute for the VM to transition to the 'running' state.
+def wait_until_vm_state(vm_uuid, target_state):
+    """Wait for VM state to change.
+
+    Return True if the state changed to the target_stated within one minute.
+    Return False otherwise.
     """
-    try:
+    timeout = 60  # seconds
+    check_interval = 5  # seconds
+    start_time = time.time()
+    while time.time() - start_time < timeout:
         vm_state = get_vm_state(vm_uuid)
-        if vm_state != "running":
-            print(
-                f"VM {vm_uuid} is not running (state: {vm_state}). Starting VM..."
-            )
-            run_vboxmanage(["startvm", vm_uuid, "--type", "gui"])
-
-            # Wait for VM to start (up to 1 minute)
-            timeout = 60  # seconds
-            check_interval = 5  # seconds
-            start_time = time.time()
-            while time.time() - start_time < timeout:
-                vm_state = get_vm_state(vm_uuid)
-                if vm_state == "running":
-                    print(f"VM {vm_uuid} started.")
-                    time.sleep(5)  # wait a bit to be careful and avoid any weird races
-                    return
-                print(f"Waiting for VM (state: {vm_state})")
-                time.sleep(check_interval)
-            print("Timeout waiting for VM to start. Exiting...")
-            raise TimeoutError(
-                f"VM did not start within the timeout period {timeout}s."
-            )
-        else:
-            print("VM is already running.")
-            return
-    except Exception as e:
-        raise Exception(f"Could not ensure '{vm_uuid}' running") from e
+        if vm_state == target_state:
+            time.sleep(5)  # wait a bit to be careful and avoid any weird races
+            return True
+        time.sleep(check_interval)
+    return False
+
+
+def ensure_vm_running(vm_uuid):
+    """Start the VM if its state is not 'running'."""
+    vm_state = get_vm_state(vm_uuid)
+    if vm_state == "running":
+        return
+
+    print(f"VM {vm_uuid} state: {vm_state}. Starting VM...")
+    run_vboxmanage(["startvm", vm_uuid, "--type", "gui"])
+
+    if not wait_until_vm_state(vm_uuid, "running"):
+        raise TimeoutError(f"Unable to start VM {vm_uuid}.")
 
 
 def ensure_vm_shutdown(vm_uuid):
-    """Checks if the VM is running and shuts it down if it is."""
-    try:
-        vm_state = get_vm_state(vm_uuid)
-        if vm_state == "saved":
-            print(
-                f"VM {vm_uuid} is in a saved state. Powering on for a while then shutting down..."
-            )
-            ensure_vm_running(vm_uuid)
-            time.sleep(120)  # 2 minutes to boot up
```

Comment:
Still need this to ensure the guest is fully booted to the desktop before initiating a powerdown.

Reply:
We are just starting the VM here because we can't shut it down in a saved state. Once the state is running, the shutdown works without any extra waiting. Or am I missing something?

I have tested this code with several saved states (saving it before it is booted to the desktop for example) and this code works for all of them:
```
./vbox-export-snapshots.py 
Starting operations on FLARE-VM
VM {2bc66f50-9ecb-4b10-a4dd-0cc329bc383d} state: saved. Starting VM...
VM {2bc66f50-9ecb-4b10-a4dd-0cc329bc383d} state: running. Shutting down VM...
```

Reply:
If there's a wait in the vbox-export-snapshots I think that would be fine. I recall from testing I did have to wait a short while for it to work well.